src/contracts/pods/EigenPod.sol

```diff
@@ -538,6 +491,78 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
+        require(
+            validatorInfo.mostRecentBalanceUpdateTimestamp < oracleTimestamp,
+            "EigenPod.verifyBalanceUpdate: Validators balance has already been updated for this timestamp"
+        );
+
+        // 2. Balance updates should only be performed on "ACTIVE" validators
+        require(
+            validatorInfo.status == VALIDATOR_STATUS.ACTIVE, 
+            "EigenPod.verifyBalanceUpdate: Validator not active"
+        );
+
+        // 4. Balance updates should only be made before a validator is fully withdrawn. 
+        // -- A withdrawable validator may not have withdrawn yet, so we require their balance is nonzero
+        // -- A fully withdrawn validator should withdraw via verifyAndProcessWithdrawals
+        if (validatorFields.getWithdrawableEpoch() <= _timestampToEpoch(oracleTimestamp)) {
+            require(
+                validatorBalance > 0,
+                "EigenPod.verifyBalanceUpdate: validator is withdrawable but has not withdrawn"
+            );
+        }
+
+        // Verify passed-in validatorFields against verified beaconStateRoot:
+        BeaconChainProofs.verifyValidatorFields({
+            beaconStateRoot: beaconStateRoot,
+            validatorFields: validatorFields,
+            validatorFieldsProof: balanceUpdateProof.validatorFieldsProof,
+            validatorIndex: validatorIndex
+        });
+
+        // Verify passed-in validator balanceRoot against verified beaconStateRoot:
+        BeaconChainProofs.verifyValidatorBalance({
+            beaconStateRoot: beaconStateRoot,
+            balanceRoot: balanceUpdateProof.balanceRoot,
+            validatorBalanceProof: balanceUpdateProof.validatorBalanceProof,
+            validatorIndex: validatorIndex
+        });
+
+        // Done with proofs! Now update the validator's balance and send to the EigenPodManager if needed
+
+        uint64 currentRestakedBalanceGwei = validatorInfo.restakedBalanceGwei;
+        uint64 newRestakedBalanceGwei = _calculateRestakedBalanceGwei(validatorBalance);
+        
+        // Update validator balance and timestamp, and save to state:
+        validatorInfo.restakedBalanceGwei = newRestakedBalanceGwei;
+        validatorInfo.mostRecentBalanceUpdateTimestamp = oracleTimestamp;
+        _validatorPubkeyHashToInfo[validatorPubkeyHash] = validatorInfo;
+
+        // If our new and old balances differ, calculate the delta and send to the EigenPodManager
+        if (newRestakedBalanceGwei != currentRestakedBalanceGwei) {
+            emit ValidatorBalanceUpdated(validatorIndex, oracleTimestamp, newRestakedBalanceGwei);
+
+            int256 sharesDeltaGwei = _calculateSharesDelta({
+                newAmountGwei: newRestakedBalanceGwei,
+                previousAmountGwei: currentRestakedBalanceGwei
+            });
+            eigenPodManager.recordBeaconChainETHBalanceUpdate(podOwner, sharesDeltaGwei * int256(GWEI_TO_WEI));
+        }
```

Comment:
shouldn't you group these together and just make a single call at the end, outside of the loop, as we now do with withdrawals?

Reply:
to elaborate, seems like the calls to `eigenPodManager.recordBeaconChainETHBalanceUpdate` could be bundled together inside the main body of `verifyBalanceUpdates` -- the internal function `_verifyBalanceUpdate` could return the sharesDeltaGwei for this.

Reply:
done